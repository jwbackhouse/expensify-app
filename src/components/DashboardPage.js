import React from 'react';
import ExpenseList from './ExpenseList'
import ExpenseListFilter from './ExpenseListFilter'

const DashboardPage = () => (
  <div>
    <p>This is the homepage</p>
    <ExpenseListFilter />
    <ExpenseList />
  </div>
);

export default DashboardPage;