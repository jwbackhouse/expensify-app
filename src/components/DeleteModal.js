import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';

// TODO:
export class DeleteModal extends React.Component {
  componentDidMount = () => {
    Modal.setAppElement('body');
  }
  
  render() {
    return (
      <Modal
        isOpen={!!this.props.showModal}
        onRequestClose={this.props.onModalCancel}   // Allows esc / clicking background to close modal
        contentLabel="Just to check..."
      >
        <h3>Are you sure you want to delete this expense?</h3>
        <button
          className='button--secondary'
          onClick={this.props.onModalConfirm}
        >Yes</button>
        <button
          className='button--secondary'
          onClick={ this.props.onModalCancel }
        >No</button>
      </Modal>
    );
  }
};

export default DeleteModal;