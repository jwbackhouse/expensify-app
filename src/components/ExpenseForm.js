import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  // Populate fields if data passed on EditPage, otherwise leave blank for AddExpensePage
  state = {
    description: this.props.expense ? this.props.expense.description : '',
    amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
    notes: this.props.expense ? this.props.expense.notes : '',
    createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
    calendarFocused: false,
    error: ''
  };
  
  // Update state.description with typed value, which then renders into the form
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  
  // Ditto for notes
  onNoteChange = (e) => {
    const notes = e.target.value;
    this.setState(() => ({ notes }));
  };
  
  // Ditto for amount
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {    // Allow no entry || Enforce max two decimal places
      this.setState(() => ({amount}));
    }
  };
  
  // Date picker functions
  onDateChange = (createdAt) => {
    if(createdAt) {
      this.setState(() => ({createdAt}));
    }   // No else clause so cannot delete date field
  };
  
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  
  onSubmit = (e) => {
    e.preventDefault();
    
    if (!this.state.description || !this.state.amount) {    // Check description and amount are present
      this.setState(() => ({error: 'Please complete both the description and amount'}));    // Error message
    } else {
      this.setState(() => ({error: ''}));   // Clear error message
      this.props.onSubmit({   // Provide expense object to props function
        description: this.state.description,
        amount: parseFloat(this.state.amount,10) * 100,    // Convert amount string to number and convert into pence
        createdAt: this.state.createdAt.valueOf(),   // Convert moment object to Unix ms
        notes: this.state.notes
      });
    }
  };
  
  // Render
  render() {
    return (
      <div>
      { this.state.error && <p>this.state.error</p> }
        <form onSubmit = { this.onSubmit }>
          <input
            type = 'text'
            placeholder = 'Description'
            value = {this.state.description}
            onChange = {this.onDescriptionChange}
            autoFocus
          />
          <input
            type = 'text'
            placeholder = 'Amount'
            value = { this.state.amount }
            onChange = { this.onAmountChange }
          />
          <SingleDatePicker
            date = { this.state.createdAt }
            onDateChange = { this.onDateChange }
            focused = { this.state.calendarFocused }    // Is the calendar open or closed
            onFocusChange = { this.onFocusChange }
            numberOfMonths = { 1 }    // # months to show
            isOutsideRange={() => false }   // Allow dates in the past
            displayFormat={() => "DD/MM/YYYY"}    // Convert to UK date format
          />
          <textarea
            type = 'text'
            value = {this.state.notes}
            placeholder = 'Enter details here'
            onChange = { this.onNoteChange }
          ></textarea>
          <button>
            { this.props.buttonText ? this.props.buttonText : 'Add expense' }
          </button>
        </form>
      </div>
    )
  }
};