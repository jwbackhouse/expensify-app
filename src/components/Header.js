import React from 'react';
import { Link, NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';


export const Header = ({startLogout}) => (
  <div>
    <h1>
      <Link to='/dashboard'>Expensify</Link>
    </h1>
    <ul>
      <li>
        <NavLink to='/dashboard' activeClassName='is-active' exact={true}>Dashboard</NavLink>
      </li>
      <li>
        <NavLink to='/create' activeClassName='is-active' exact={true}>Create</NavLink>
      </li>
      <li>
        <NavLink to='/help' activeClassName='is-active' exact={true}>Help</NavLink>
      </li>
    </ul>
    <button onClick={ startLogout }>Logout</button>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);