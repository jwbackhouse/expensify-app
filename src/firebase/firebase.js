import * as firebase from 'firebase';

const firebaseConfig = {
  apiKey: "AIzaSyAMlaXLBe2QL3E52M7B5vAxBmurHpiFZWo",
  authDomain: "expensify-dfa83.firebaseapp.com",
  databaseURL: "https://expensify-dfa83.firebaseio.com",
  projectId: "expensify-dfa83",
  storageBucket: "expensify-dfa83.appspot.com",
  messagingSenderId: "1089954814599",
  appId: "1:1089954814599:web:7b87ed1047c0d315efb622",
  measurementId: "G-HGYB92KNHS"
};

firebase.initializeApp(firebaseConfig);
// TODO: link analytics functionality
// firebase.analytics();

const database = firebase.database();

database.ref().remove();

const expenses = [{
  description: 'Expense One',
  body: 'This is the body',
  amount: 500,
  createdAt: '1234'
},{
  description: 'Expense Two',
  body: 'This is the body',
  amount: 600,
  createdAt: '5678'
},{
  description: 'Expense Three',
  body: 'This is the body',
  amount: 200,
  createdAt: '9101'
}];

// NOTE .push is used to store an array, with each item stored under a unique identifier
database.ref('expenses').push(expenses[0]);
database.ref('expenses').push(expenses[1]);
database.ref('expenses').push(expenses[2]);


database.ref('expenses').on('child_changed', (snapshot) => {
  console.log(snapshot.key, snapshot.val());
});

// NOTE:  How to convert Firebase output into an array, using the key as unique identifier
// database.ref('expenses').on('value', (snapshot) => {
//   const expenses = [];
//   snapshot.forEach((childSnapshot) => {
//     const output = childSnapshot.val();
//     expenses.push({
//       id: childSnapshot.key,
//       ...output
//     })
//   })
//   console.log(expenses);
//   });



// NOTE: How to subscribe to automatic updates (does not return a promise and haven't worked out how to pass data out of the callback function)
// database.ref().on('value', (snapshot) => {
//   const person = snapshot.val();
//   console.log(`${ person.name } is ${ person.age } years old and comes from ${ person.location.country }`);
// })

// NOTE: How to make one-off data pull (returns a promise)
// database.ref('expenses')
//   .once('value')
//   .then((snapshot) => {
//     console.log(snapshot.val())
//   });

// NOTE: How to set data
// database.ref().set({
//   name: 'James Backhouse',
//   age: 38,
//   location: {
//     country: 'UK',
//     county: 'Surrey'
//   }
// }).then(() => {
//   console.log('Data is saved');
// }).catch((e) => {
//   console.log('This failed.', e);
// });

// NOTE: How to set data for specific field
// database.ref('attributes').set({
//   height: 180,
//   weight: 73
// });

// NOTE: How to remove specific field
// database.ref('age').remove().then(() => {
//   console.log('Removed')
// }).catch((e) => {
//   console.log('Error:', e)
// });

// NOTE: How to add / edit / remove using update
// database.ref().update({
//   name:'Florence Backhouse',
//   age: 4,
//   favouriteColour: 'Purple',
//   'attributes/height': 40,    // NOTE: how to get to nested entries
//   'attributes/weight': null   // NOTE: how to delete entries
// })