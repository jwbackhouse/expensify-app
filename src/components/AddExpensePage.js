import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);    // Add expense to state under user's account
    this.props.history.push('/');    // Redirect on submit (uses in-built method)
  };
  
  render() {
    return (
      <div>
        <p>This is the add expense page</p>
        <ExpenseForm
          onSubmit={ this.onSubmit }
        />
      </div>
    );
  };
};

// Create bespoke dispatch method in order to simplify the above (and make it possible to test with a spy)
const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);    // first arg is state, second is dispatch


// *** Previous in-line version of the above (which couldn't be tested in the same way)
// const AddExpensePage = (props) => (
//   <div>
//     <p>This is the add expense page</p>
//     <ExpenseForm onSubmit={(expense) => {
//         props.dispatch(addExpense(expense));    // Add expense to state
//         props.history.push('/');    // Redirect on submit (uses in-built method)
//       }}
//     />
//   </div>
// );