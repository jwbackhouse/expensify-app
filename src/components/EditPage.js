import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import DeleteModal from './DeleteModal';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditPage extends React.Component {
  state = {
    showModal: false
  };
  
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);    // Edit expense
    this.props.history.push('/');   // Redirect on submit (uses in-built method)
  };
  
  onClickDelete = () => {
    this.setState(() => ({ showModal: true }));
  };
  
  onClickCancel = () => {
    this.props.history.push('/');
  };
  
  onModalConfirm = () => {
    this.props.startRemoveExpense(this.props.expense.id);    // Delete expense
    this.setState(() => ({showModal: false}));
    this.props.history.push('/');   // Redirect on submit (uses in-built method)
  };
  
  onModalCancel = () => {
    this.setState(() => ({showModal: false}));
  };
  
  render() {
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Edit expense</h1>
          </div>
        </div>
        <div className='content-container'>
          <ExpenseForm
            expense = {this.props.expense}
            buttonText = { 'Save changes' }
            onSubmit = { this.onSubmit }
          />
          <div>
            <button
              className='button button--pink button--margin-right'
              id='deleteButton'
              onClick = { this.onClickDelete }
            >Delete</button>
            <button
              className='button button--grey'
              id='cancelButton'
              onClick = { this.onClickCancel }
            >Cancel</button>
          </div>
        </div>
        <DeleteModal onModalConfirm={ this.onModalConfirm } onModalCancel={ this.onModalCancel } showModal={this.state.showModal} />
      </div>
    );
  };
  
  // *** PREVIOUS CODE BEFORE MAPTODISPATCH ADDED
  // return (
  //   <div>
  //     <ExpenseForm
  //       expense = {props.expense}
  //       buttonText = { 'Save changes' }
  //       onSubmit = {(expense) => {
  //         props.dispatch(editExpense(props.expense.id, expense));    // Edit expense
  //         props.history.push('/');    // Redirect on submit (uses in-built method)
  //       }}
  //     />
  //     <button onClick = {() => {
  //       props.dispatch(removeExpense({ id: props.expense.id }));    // Delete expense
  //       props.history.push('/');
  //     }}>Delete</button>
  //   </div>
  // );
}

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense({ id }))
})

const mapStateToProps= (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)   // Matches ID pulled from URL
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPage);