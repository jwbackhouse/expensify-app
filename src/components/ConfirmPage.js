import React from 'react';
import { Link } from 'react-router-dom';
import {firebase} from '../firebase/firebase';
import Header from './Header';


export class ConfirmPage extends React.Component {
  state = {
    email: '',
    error: '',
    showForm: false
  };
  
  componentDidMount = () => {
    this.isValidURL();
  }
  
  onEmailChange = (e) => {
    const email = e.target.value;
    this.setState(() => ({ email }));
  };

  onSubmit = (e) => {
    e.preventDefault();
    this.signIn(this.state.email);
  };
  
  signIn = (email) => {
    firebase.auth().signInWithEmailLink(email, window.location.href)    // auth code parsed from URL
      .then((result) => {
        window.localStorage.removeItem('emailForSignIn');   // Clear email from storage.
        // Can access the new user via result.user
      })
      .catch((err) => {
        const error = 'Sorry, something went wrong.'
        this.setState(() => ({error}))
      });
  };
  
  isValidURL = () => {
    const pageURL = firebase.auth().isSignInWithEmailLink(window.location.href);
    const savedEmail = window.localStorage.getItem('emailForSignIn');
    if (pageURL) {    // Check page URL contains valid auth code
      if (!savedEmail) {    // If local storage doesn't contain an email (e.g. user switched devices)
        const showForm = true;
        this.setState(() => ({showForm}))
      } else {
        this.signIn(savedEmail);
      }
    } else {
      const error = "Sorry, that doesn't look like a valid link."
      this.setState(() => ({error}))
    }
    
  }
  
  onClick = () => {
    this.isValidURL();
  }

  render() {
    let form, error;
    if (this.state.showForm) {
      form = (
        <div>
          <h3>Please confirm your email</h3>
          <form onSubmit = { this.onSubmit }>
            <input
              type='email'
              placeholder='Email address'
              value={this.state.email}
              onChange={this.onEmailChange}
            />
            <button className="button--tertiary-pink">Confirm</button>
          </form>
        </div>
      )
    }
    if (this.state.error) {
      error = (
        <div>
          <p>{ this.state.error }</p>
          <Link className='button' to='/'>Return to login page</Link>
        </div>
      )
    }
    return (
      <div className='box-layout'>
        <div className='box-layout__box'>
          <h1 className='box-layout__title'>Expensify</h1>
          { form }
          { error }
        </div>
      </div>
    )
  }
};

export default ConfirmPage;