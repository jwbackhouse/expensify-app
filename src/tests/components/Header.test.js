import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';
import { firebase } from '../../firebase/firebase';

let startLogout, wrapper;
beforeEach(() => {
  startLogout = jest.fn();
  wrapper = shallow(<Header startLogout={startLogout} />);
});

test('Should render Header correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should call the startLogout action when logout button clicked', () =>{
  wrapper.find('button').simulate('click');
  expect(startLogout).toBeCalled();
  // expect(firebase.auth().currentUser).toEqual(null);
});



// *** CAN AVOID NEEDING TO IMPORT TOJSON BY ADDING "snapshotSerializers": ["enzyme-to-json/serializer"] TO JEST.CONFIG.JSON
// import toJSON from 'enzyme-to-json';
// test('Should render Header correctly', () => {
//   const wrapper = shallow(<Header />);
//   expect(toJSON(wrapper)).toMatchSnapshot;
// })


// *** PREVIOUS APPROACH USING BASIC RENDERER (BETTER TO USE ENZYME AS ABOVE) ***
// import ReactShallowRenderer from 'react-test-renderer/shallow';
// test('Should render Header correctly', () => {
//   const renderer = new ReactShallowRenderer();
//   renderer.render(<Header />);
//   expect(renderer.getRenderOutput()).toMatchSnapshot();
// })