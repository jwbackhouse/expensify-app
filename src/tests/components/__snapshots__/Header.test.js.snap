// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Header correctly 1`] = `
<div>
  <h1>
    <[object Object]
      to="/dashboard"
    >
      Expensify
    </[object Object]>
  </h1>
  <ul>
    <li>
      <[object Object]
        activeClassName="is-active"
        exact={true}
        to="/dashboard"
      >
        Dashboard
      </[object Object]>
    </li>
    <li>
      <[object Object]
        activeClassName="is-active"
        exact={true}
        to="/create"
      >
        Create
      </[object Object]>
    </li>
    <li>
      <[object Object]
        activeClassName="is-active"
        exact={true}
        to="/help"
      >
        Help
      </[object Object]>
    </li>
  </ul>
  <button
    onClick={[MockFunction]}
  >
    Logout
  </button>
</div>
`;
